### PR TITLE
ARI Beamline Simulation based on XRT

### PR DESCRIPTION
**Pull request by Jiemin-Li**
This is the first working version of ARI M1-section simulation based on XRT. The XRT is a python package that can perform the ray tracking for beamline, i.e., simulating the beam properties as propagating through different components in beamline. There are two files in xrt_sim: one is the initialization of ARI M1 parameters and the other one is to build up the AriModel based on the input parameters. The resulting AriModel includes all beamline components (such as source, m1, m2, slit, screen...) that wrap up the beam properties (such as intensity, energy, polarization...). The next step will be integrating this simulation with the developed Caproto IOCs to mimic the beamline. @awalter-bnl and @tfg4280  please have a look and merge it to the origin repo once you have reviewed it.

### CODE REVIEW

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 1]
```diff
@@ -0,0 +1,674 @@
+import numpy as np
+# import matplotlib.pyplot as plt
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: Can you remove these unused import statements here and one line below?

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+import xrt.backends.raycing as xrt_raycing
+import xrt.backends.raycing.sources as xrt_source
+import xrt.backends.raycing.apertures as xrt_aperture
+import xrt.backends.raycing.screens as xrt_screen
+import xrt.backends.raycing.oes as xrt_oes
+# import xrt.backends.raycing.materials as xrt_material
+
+from bl_initialization import BLparams
+
+#################################################################
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: This line also should be removed

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+        -------
+        updated : Boolean
+            Potentially modified the input parameter updated if the update
+            indicates a re-activation required.
+        '''
+
+        if update != None:
+            for pv_name, pv_val in update.items():
+                if pv_name.split(':')[1] in ['x', 'y', 'z']:
+                    center_list = getattr(self, 'center').copy() # self.pv2xrt[pv_name.split(':')[0]][pv_name]
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: Is the comment at the end here still needed?

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+        as the beamline optics has been modified.
+
+        Returns
+        -------
+        updated : Boolean
+            Potentially modified the input parameter updated if the update
+            indicates a re-activation required.
+        '''
+
+        # update = {'ARI_M2:x': 0.2}
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: Another commented out code line that we should remove

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+        if update != None:
+            # update = {'ARI_M2:x': 0.2, 'ARI_M1:Ry_fine': 3.1}
+            for pv_name, pv_val in update.items():
+                if pv_name.split(':')[1] in ['x', 'y', 'z']:
+                    center_list = getattr(self, 'center').copy() # self.pv2xrt[pv_name.split(':')[0]][pv_name]
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: Again here the comment at the end is probably not needed

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+                    for i, p in enumerate(['x','y','z']):
+                        if pv_name.split(':')[1] == p:
+                            unit_vector_trans_NSLS2[i] = 1.0
+                    unit_vector_trans_XRT = np.dot(self.coordinate_NSLS2XRT[self.deflection],
+                                                   unit_vector_trans_NSLS2)
+
+                    center_list[int(np.where(unit_vector_trans_XRT!=0)[0][0])] = (
+                            pv_val * unit_vector_trans_XRT[int(np.where(unit_vector_trans_XRT!=0)[0][0])])
+                    if getattr(self, 'center') != center_list:
+                        setattr(self, 'center', center_list) # self.pv2xrt[pv_name.split(':')[0]][pv_name]
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: Please consider removing the comment at the end.

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+        as the beamline aperature has been modified.
+
+        Returns
+        -------
+        updated : Boolean
+            Potentially modified the input parameter updated if the update
+            indicates a re-activation required.
+
+        '''
+        # update = {'ARI_M1:diag:outboard':2.3}
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: Once more can you remove this comment line

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+        if update != None:
+            for pv_name, pv_val in update.items():
+                if pv_name.split(':')[-1] in ['x', 'y', 'z']:
+                    center_list = getattr(self, 'center').copy() # self.pv2xrt[pv_name.split(':')[0]][pv_name]
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: and again is the comment at the end required?

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 9]
```diff
@@ -0,0 +1,674 @@
...
+                    for i, p in enumerate(['x','y','z']):
+                        if pv_name.split(':')[-1] == p:
+                            unit_vector_trans_NSLS2[i] = 1.0
+                    unit_vector_trans_XRT = np.dot(self.coordinate_NSLS2XRT['upward'],
+                                                   unit_vector_trans_NSLS2)
+
+                    center_list[int(np.where(unit_vector_trans_XRT!=0)[0][0])] = (
+                            pv_val * unit_vector_trans_XRT[int(np.where(unit_vector_trans_XRT!=0)[0][0])])
+                    if getattr(self, 'center') != center_list:
+                        setattr(self, 'center', center_list) # self.pv2xrt[pv_name.split(':')[0]][pv_name]
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: comment at the end should be removed

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 10]
```diff
@@ -0,0 +1,674 @@
...
+        updated : Boolean
+            Potentially modified the input parameter updated if the update
+            indicates a re-activation required.
+
+        '''
+
+        if update != None:
+            for pv_name, pv_val in update.items():
+                if pv_name.split(':')[-1] in ['x', 'y', 'z']:
+                    center_list = getattr(self, 'center').copy() # self.pv2xrt[pv_name.split(':')[0]][pv_name]
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: please remove the comment at the end of the line

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 11]
```diff
@@ -0,0 +1,674 @@
...
+                    for i, p in enumerate(['x','y','z']):
+                        if pv_name.split(':')[-1] == p:
+                            unit_vector_trans_NSLS2[i] = 1.0
+                    unit_vector_trans_XRT = np.dot(self.coordinate_NSLS2XRT['upward'],
+                                                   unit_vector_trans_NSLS2)
+
+                    center_list[int(np.where(unit_vector_trans_XRT!=0)[0][0])] = (
+                            pv_val * unit_vector_trans_XRT[int(np.where(unit_vector_trans_XRT!=0)[0][0])])
+                    if getattr(self, 'center') != center_list:
+                        setattr(self, 'center', center_list) # self.pv2xrt[pv_name.split(':')[0]][pv_name]
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: again please consider removing the comment at the end

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 12]
```diff
@@ -0,0 +1,674 @@
...
+    Methods
+    -------
+    *methods : many
+
+    """
+
+    def __init__(self, update_comp=None):
+        self.update_comp = update_comp
+
+    ###############################################
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: I prefer to not have these comment lines can you remove them here and the two below?

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 13]
```diff
@@ -0,0 +1,674 @@
...
+                        disty=None, dy=0,
+                        distz='normal', dz=0.001,
+                        distxprime='normal', dxprime=0.0001,  # source angular profile
+                        distzprime='normal', dzprime=0.01,
+                        distE='normal', energies=(energy_ref, energy_sigma),  # source energy profile
+                        polarization='horizontal',
+                        filamentBeam=False,
+                        uniformRayDensity=False)
+    source.activate(updated=True)
+    ###############################################
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: Again consider removing these comment lines here and a few lines below

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 14]
```diff
@@ -0,0 +1,674 @@
...
+    screen1 = ID29Screen(bl=bl,
+                         name='M1Diag_Scrn',
+                         center=blG['rSrcM1Diag'],
+                         x=np.array([1, 0, 0]),
+                         z=np.array([0, 0, 1]),
+                         upstream=slit2)
+    screen1.activate(updated=True)
+
+
+    #
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: ALl of these commented out lines below can be removed

File: src/ari_sxn_simbeamline/xrt_sim/bl_initialization.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    1. blG is a dictionary containing all parameters for beamline simulation.
+
+    2. update_para is a dictionary containing the parameters that are modified
+    comparing to their initial values. update_para should be a sub-dict of blG.
+
+    Note: The distance is in the unit of mm,
+        and the angle is in the unit of radian.
+    '''
+    blG = dict()
+    ############################################################################
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/bl_initialization.py]: Again I would prefer not to include these comment lines here and below

**Review by awalter-bnl — CHANGES_REQUESTED** [review]
Hi @Jiemin-Li , Thanks for the effort in getting this PR in. It is a lot of work and overall looks good. I have made a few line comments, asking to remove some commented out code lines and some other comment lines that I am not sure are required. Once that is done this is ready to merge IMO.

File: src/ari_sxn_simbeamline/xrt_sim/bl_initialization.py (hunk 15 above)
Comment (Jiemin-Li) [reply src/ari_sxn_simbeamline/xrt_sim/bl_initialization.py]: @awalter-bnl All raised issues have been resolved, it should be ready to be merged.

File: src/ari_sxn_simbeamline/xrt_sim/ari_sim.py [hunk 16]
```diff
@@ -0,0 +1,668 @@
...
+    screen1 = ID29Screen(bl=bl,
+                         name='M1Diag_Scrn',
+                         center=blG['rSrcM1Diag'],
+                         x=np.array([1, 0, 0]),
+                         z=np.array([0, 0, 1]),
+                         upstream=slit2)
+    screen1.activate(updated=True)
+
+
+    # The update method needs to be constructed during the integration with Caproto IOC.
```
Comment (awalter-bnl) [inline src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: @Jiemin-Li , Sorry to push on this again but can you remove the commented out 'update' function below or un comment it if it is needed?

Comment (Jiemin-Li) [reply src/ari_sxn_simbeamline/xrt_sim/ari_sim.py]: @awalter-bnl, just remove the update method, but we may need to build one similar method when integrating with the Caproto IOCs in next step.

**Comment by awalter-bnl** [comment]
Thanks for removing the code @Jiemin-Li , I agree we will need something similar when we integrate but that should be left for that PR !-).

**Review by awalter-bnl — APPROVED** [review]

**Pull request merged**